braindecode/datautil/windowers.py

```diff
@@ -66,95 +45,145 @@ class EventWindower(Windower):
 
-class FixedLengthWindower(Windower):
-    """
-    A Windower that creates supercrops/windows based on fake events that equally
-    divide the continuous signal.
+
+def create_fixed_length_windows(
+        concat_ds, start_offset_samples, stop_offset_samples,
+        supercrop_size_samples, supercrop_stride_samples, drop_samples,
+        mapping=None):
+    """Windower that creates sliding supercrops/windows.
 
     Parameters
     ----------
-    trial_start_offset_samples: int
-        start offset from original trial onsets in samples
-    trial_stop_offset_samples: int
-        stop offset from original trial onsets in samples
+    concat_ds: ConcatDataset
+        a concat of base datasets each holding raw and descpription
+    start_offset_samples: int
+        start offset from beginning of recording in samples
+    stop_offset_samples: int
+        stop offset from beginning of recording in samples.
     supercrop_size_samples: int
         supercrop size
     supercrop_stride_samples: int
         stride between supercrops
     drop_samples: bool
         whether or not have a last overlapping supercrop/window, when
-        supercrops/windows do not equally devide the continuous signal
+        supercrops/windows do not equally divide the continuous signal
     mapping: dict(str: int)
         mapping from event description to target value
+
+    Returns
+    -------
+    windows_ds: WindowsDataset
+        Dataset containing the extracted windows.
     """
-    def __call__(self, base_ds):
+    _check_windowing_arguments(
+        start_offset_samples, stop_offset_samples,
+        supercrop_size_samples, supercrop_stride_samples)
+
+    list_of_windows_ds = []
+    for ds in concat_ds.datasets:
         # already includes last incomplete supercrop start
-        starts = np.arange(0, base_ds.raw.n_times, self.stride)
-        # 1/0
-        if self.drop_samples:
+        stop = (ds.raw.n_times
+                if stop_offset_samples is None
+                else stop_offset_samples)
+        starts = np.arange(
+            start_offset_samples, stop, supercrop_stride_samples)
+
+        last_allowed_ind = stop - supercrop_size_samples
+
+        if starts[-1] > last_allowed_ind:
             starts = starts[:-1]
-        else:
+        if not drop_samples:
             # if last supercrop does not end at trial stop, make it stop
             # there
-            if starts[-1] != base_ds.raw.n_times - self.size:
-                starts[-1] = base_ds.raw.n_times - self.size
+            if starts[-1] < last_allowed_ind:
+                starts[-1] = last_allowed_ind
 
-        # TODO: handle multi-target case
-        assert len(base_ds.info[base_ds.target]) == 1, (
-            "multi-target not supported")
-        description = base_ds.info[base_ds.target].iloc[0]
+        # TODO: handle multi-target case / non-integer target case
+        if ds.target is None:
+            target = 1  # default target
```

Comment:
What was the idea behind this default target here @hubertjb ? I think this is critical and should not be. It might cause running classification on wrong targets without the user even noticing. I think we should raise an error here, when the target is None.

Reply:
I see your point, we don't want to create fake targets that a user could end up using by mistake. The reason I included a default target is I believe there are some situations where you don't actually want to specify a target here: 1) you're not going to use a target in a supervised classification sense (e.g., structured prediction tasks, autoencoding, etc.) or 2) the target will be created by the sampler itself later (e.g., in a contrastive learning task where the sampler picks two windows and provides a target of 0 if they are far away or 1 if they are close).

One solution would be to have the default target be a type that couldn't be used as a target later on, and so an unsuspecting user would get an error when training. What do you think?

Reply:
If you want a type that cannot be used as target, None would work. However, in the create_windows_from_events we explicitly require an integer target right now: https://github.com/braindecode/braindecode/blob/1bbe5fe376664330a4dedbd58509f37c8d18c8d5/braindecode/datautil/windowers.py#L159-L163

We could do it like this:
```
target = ds.target
if mapping is not None, ("a mapping from '{target}' to int is required"):
    target = mapping[target]
```

Reply:
Ok cool, I think that would work!

Reply:
@gemeinl actually that didn't work, because MNE requires an int in the third column of the event array. One hacky-ish solution would be to use an int (e.g., 0) in the third column of the event array when calling mne.Epochs(), but actually saving None in the metadata object. What do you think?

Reply:
Ah, yes, I see. Maybe we could go with -1? I think you already did this in 4613b63. And yes, we should save None in the metadata. Eventually, the target is read from there: https://github.com/braindecode/braindecode/blob/4613b639976b6330c16ced433fd46a72c5cfdd08/braindecode/datasets/base.py#L65